### PR TITLE
feat: add ERC2981 component

### LINKED ISSUE #1041: ERC2981 implementation in Cairo
**🧐 Motivation**
[ERC2981](https://eips.ethereum.org/EIPS/eip-2981) is a standardized way to retrieve royalty payment information for non-fungible tokens (NFTs) to enable universal support for royalty payments across all NFT marketplaces and ecosystem participants.

**📝 Details**
There are many marketplaces for NFTs with multiple unique royalty payment implementations that are not easily compatible or usable by other marketplaces. Just like the early days of ERC-20 tokens, NFT marketplace smart contracts are varied by ecosystem and not standardized. ERC2981 enables all marketplaces to retrieve royalty payment information for a given NFT. This enables accurate royalty payments regardless of which marketplace the NFT is sold or re-sold at.

Here is a proposal for ERC2981 interface:
```cairo
const IERC2981_ID: felt252 = 0x2d3414e45a8700c29f119a54b9f11dca0e29e06ddcb214018fc37340e165ed6;

#[starknet::interface]
pub trait IERC2981<TState> {
    fn royalty_info(self: @TState, token_id: u256, sale_price: u256) -> (ContractAddress, u256);
}
```

Please note that I have wrote [a first implementation](https://github.com/ArkProjectNFTs/ark-project/tree/acda11aa6f25cf18146993148cff6f6079af10da/contracts/ark_oz) for the [Ark Project](https://www.arkproject.dev/), I can adapt for addition to OZ Cairo library.

### PR DESCRIPTION
**Pull request by ptisserand**
Fixes #1041 

This PR add [`ERC2981`](https://eips.ethereum.org/EIPS/eip-2981) component.

ERC2981 component implements the `ERC2981` standard interface:
```cairo
pub const IERC2981_ID: felt252 = 0x2d3414e45a8700c29f119a54b9f11dca0e29e06ddcb214018fc37340e165ed6;

#[starknet::interface]
pub trait IERC2981<TState> {
    fn royalty_info(self: @TState, token_id: u256, sale_price: u256) -> (ContractAddress, u256);
}
```

The following internal functions are implemented:
```cairo
fn initializer(ref self: TState, default_receiver: ContractAddress, default_royalty_fraction: u256);
fn _fee_denominator(self: @TState) -> u256;
fn _default_royalty(self: @TState) -> (ContractAddress, u256, u256);
fn _set_default_royalty(ref self: TState, receiver: ContractAddress,fee_numerator: u256,); 
fn _set_token_royalty(ref self: TState, token_id: u256, receiver: ContractAddress, fee_numerator: u256); 
fn _token_royalty(self: @TState, token_id: u256) -> (ContractAddress, u256, u256);
fn _reset_token_royalty(ref self: TState, token_id: u256);
```

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [ ] Added entry to CHANGELOG.md 
- [ ] Tried the feature on a public network

### CODE REVIEW

File: src/token.cairo [hunk 1]
```diff
@@ -1,3 +1,4 @@
 pub mod erc1155;
 pub mod erc20;
+pub mod erc2981;
```
Comment (ericnordelo) [inline src/token.cairo:3]: Since this is not a different token but an extension to both ERC721 and ERC1155, it makes sense for it to be placed under a common module instead.

File: src/token/erc2981.cairo [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+pub mod interface;
+
+pub use erc2981::ERC2981Component;
+pub use fees::FeesImpl;
+pub use fees::FeesRatio;
+pub use fees::FeesRatioDefault;
+pub use interface::IERC2981Dispatcher;
+pub use interface::IERC2981DispatcherTrait;
+pub use interface::IERC2981SetupDispatcher;
+pub use interface::IERC2981SetupDispatcherTrait;
```
Comment (ericnordelo) [inline src/token/erc2981.cairo:12]: These imports can be consolidated into three lines.

File: src/token/erc2981/erc2981.cairo [hunk 3]
```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: MIT
```
Comment (ericnordelo) [inline src/token/erc2981/erc2981.cairo:1]: While the idea is good, we try to be as consistent as possible with modules that we have already implemented in Solidity, and we have an implementation [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/05f218fb6617932e56bf5388c3b389c3028a7b73/contracts/token/common/ERC2981.sol).

File: src/token/erc2981/interface.cairo [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+pub const IERC2981_ID: felt252 = 0x2d3414e45a8700c29f119a54b9f11dca0e29e06ddcb214018fc37340e165ed6;
+
+#[starknet::interface]
+pub trait IERC2981<TState> {
+    fn royalty_info(self: @TState, token_id: u256, sale_price: u256) -> (ContractAddress, u256);
+}
+
+
+#[starknet::interface]
+pub trait IERC2981Setup<TState> {
```
Comment (ericnordelo) [inline src/token/erc2981/interface.cairo:14]: This should be provided as internal functions, but since this interface is not part of the standard, we shouldn't provide an external implementation.

**Review by ericnordelo — CHANGES_REQUESTED** [review]
Thanks again for taking the time @ptisserand. Left some comments. The main suggestion is to be as consistent as possible with the implementation we have in Solidity.

File: src/token/erc2981/erc2981.cairo (hunk 3 above)
Comment (ptisserand) [reply src/token/erc2981/erc2981.cairo:1]: Do you mean that I should removed my `FeesRatio` structure and use `_feeDenominator` in storage as in solidity contract ?

Comment (ericnordelo) [reply src/token/erc2981/erc2981.cairo:1]: The solidity implementation doesn't have `_feeDenominator` in storage, but as part of the bytecode as a function. I mean to be as consistent as possible, only diverging when the benefit is obvious or there is a feature inconsistency.

File: src/token.cairo (hunk 1 above)
Comment (ptisserand) [reply src/token.cairo]: I have moved the implementation to `src/token/common/erc2981`

File: src/token/erc2981/erc2981.cairo (hunk 3 above)
Comment (ptisserand) [reply src/token/erc2981/erc2981.cairo]: Thanks for your explanation, I have updated source code to be more consistent with solidity implementation.

File: src/token/erc2981/interface.cairo (hunk 4 above)
Comment (ptisserand) [reply src/token/erc2981/interface.cairo]: I have removed this trait and added internal functions matching the ones in Solidity implementation

**Comment by ptisserand** [comment]
branch rebased on main

**Comment by ptisserand** [comment]
Hi @ericnordelo , are you agree that ERC2981 API documentation should be in `docs/modules/ROOT/pages/api/erc2981.adoc` ?
Is there a tool to generate documentation from cairo source code ?

File: src/token/common/erc2981/erc2981.cairo [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        ) -> (ContractAddress, u256) {
+            let royalty_info: RoyaltyInfo = self.token_royalty_info.read(token_id);
+            let mut royalty_receiver = royalty_info.receiver;
+            let mut royalty_fraction = royalty_info.royalty_fraction;
+
+            if royalty_receiver.is_zero() {
+                let default_royalty_info: RoyaltyInfo = self.default_royalty_info.read();
+                royalty_receiver = default_royalty_info.receiver;
+                royalty_fraction = default_royalty_info.royalty_fraction;
+            }
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:53]:
```
let (royalty_receiver, royalty_fraction) = if royalty_info.receiver.is_zero() { 
  let default_info = self.default_royalty_info.read();
  (default_info.receiver, default_info.royalty_fraction)
} else {
  (royalty_info.receiver, royalty_info.royalty_fraction)
}
```

File: src/token/common/erc2981/erc2981.cairo [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+            self._set_default_royalty(default_receiver, default_royalty_fraction)
+        }
+
+
+        /// The denominator with which to interpret the fee set in {_set_token_royalty} and
+        /// {_set_default_royalty} as a fraction of the sale price.
+        /// Defaults to 10000 so fees are expressed in basis points
+        fn _fee_denominator(self: @ComponentState<TContractState>) -> u256 {
+            10000
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:89]:
```suggestion
            10_000
```

Also makes sense to move this value to a constant

File: src/token/common/erc2981/erc2981.cairo [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        /// Returns the royalty information that all ids in this contract will default to.
+        fn _default_royalty(
+            self: @ComponentState<TContractState>
+        ) -> (ContractAddress, u256, u256) {
+            let royalty_info: RoyaltyInfo = self.default_royalty_info.read();
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:96]:
```suggestion
            let royalty_info = self.default_royalty_info.read();
```

File: src/token/common/erc2981/erc2981.cairo [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        /// - `receiver` cannot be the zero address.
+        /// - `fee_numerator` cannot be greater than the fee denominator.
+        fn _set_default_royalty(
+            ref self: ComponentState<TContractState>,
+            receiver: ContractAddress,
+            fee_numerator: u256,
+        ) {
+            let denominator = self._fee_denominator();
+            assert!(fee_numerator <= denominator, "Invalid default royalty");
+            assert!(!receiver.is_zero(), "Invalid default royalty receiver");
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:113]:
```suggestion
            assert!(receiver.is_non_zero(), "Invalid default royalty receiver");
```

File: src/token/common/erc2981/erc2981.cairo [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        ) -> (ContractAddress, u256, u256) {
+            let royalty_info: RoyaltyInfo = self.token_royalty_info.read(token_id);
+            let mut receiver = royalty_info.receiver;
+            let mut royalty_fraction = royalty_info.royalty_fraction;
+
+            if receiver.is_zero() {
+                let default_royalty_info: RoyaltyInfo = self.default_royalty_info.read();
+                receiver = default_royalty_info.receiver;
+                royalty_fraction = default_royalty_info.royalty_fraction;
+            };
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:153]: Same thing here, no need to have mutable variables, if-else loop should be preferred

File: src/tests/token/erc2981/test_erc2981.cairo [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+// 0.5% (default denominator is 10000)
+fn DEFAULT_FEE_NUMERATOR() -> u256 {
+    50
+}
+
+// 5% (default denominator is 10000)
+fn FEE_NUMERATOR() -> u256 {
+    500
+}
```
Comment (immrsd) [inline src/tests/token/erc2981/test_erc2981.cairo:38]: u256 type constants (`DEFAULT_FEE_NUMERATOR` and `FEE_NUMERATOR`) can be declared as constants, no need for function-based approach

File: src/tests/token/erc2981/test_erc2981.cairo [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    contract_address_const::<'OWNER'>()
+}
+
+fn DEFAULT_RECEIVER() -> ContractAddress {
+    contract_address_const::<'DEFAULT_RECEIVER'>()
+}
+
+fn RECEIVER() -> ContractAddress {
+    contract_address_const::<'RECEIVER'>()
+}
```
Comment (immrsd) [inline src/tests/token/erc2981/test_erc2981.cairo:28]: The constants module already declares similar constants, you can employ them instead

File: src/token/common/erc2981/erc2981.cairo [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        +HasComponent<TContractState>,
+        impl SRC5: SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>,
+    > of IERC2981<ComponentState<TContractState>> {
+        /// Returns the receiver address and amount to send for royalty
+        /// for token id `token_id` and sale price `sale_price`
+        fn royalty_info(
+            self: @ComponentState<TContractState>, token_id: u256, sale_price: u256
+        ) -> (ContractAddress, u256) {
+            let royalty_info: RoyaltyInfo = self.token_royalty_info.read(token_id);
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:45]:
```suggestion
            let royalty_info = self.token_royalty_info.read(token_id);
```

File: src/tests/token/erc2981/test_erc2981.cairo [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    assert_eq!(amount, 5000, "Wrong royalty amount");
+    let (receiver, amount) = state.royalty_info(token_id, sale_price);
+    assert_eq!(receiver, RECEIVER(), "Default receiver incorrect");
+    assert_eq!(amount, 50000, "Wrong royalty amount");
+
+    state._reset_token_royalty(token_id);
+    let (receiver, amount) = state.royalty_info(token_id, sale_price);
+    assert_eq!(receiver, DEFAULT_RECEIVER(), "Default receiver incorrect");
+    assert_eq!(amount, 5000, "Wrong royalty amount");
+}
```
Comment (immrsd) [inline src/tests/token/erc2981/test_erc2981.cairo:89]: It would be helpful to have more test cases, especially for various edge cases. For example:
- call `_set_token_royalty`, then call `_set_token_royalty` again to other values
- call `_set_token_royalty ` with zero receiver
- call `_set_token_royalty ` with zero royalty fraction
- call `_set_default_royalty ` with zero receiver
- call `_set_default_royalty ` with zero royalty fraction
- check `IERC2981_ID` is registered

**Review by immrsd — CHANGES_REQUESTED** [review]

File: src/tests/mocks/erc2981_mocks.cairo [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        ERC2981Event: ERC2981Component::Event,
+        #[flat]
+        SRC5Event: SRC5Component::Event
+    }
+
+
+    #[abi(embed_v0)]
```
Comment (ericnordelo) [inline src/tests/mocks/erc2981_mocks.cairo:28]: This should be on top of the Storage struct for consistency with the library.

**Review by ericnordelo — COMMENTED** [review]
Hey @ptisserand, the PR is looking good and we have left a few comments. We have two big migrations ongoing (as you can see [here](https://github.com/OpenZeppelin/cairo-contracts/pull/1065)), and that's why this won't make it into the next release, since those should be merged into it first. With this said, we will prioritize it and get it merged right after that release, and it will be included in the next one as the main priority.

File: src/token/common/erc2981/erc2981.cairo (hunk 5 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo]: Thanks! I forgot about this pattern.

File: src/token/common/erc2981/erc2981.cairo (hunk 6 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo]: Code updated

File: src/token/common/erc2981/erc2981.cairo (hunk 8 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo]: Code updated

File: src/token/common/erc2981/erc2981.cairo (hunk 9 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo:153]: Code updated

File: src/tests/token/erc2981/test_erc2981.cairo (hunk 10 above)
Comment (ptisserand) [reply src/tests/token/erc2981/test_erc2981.cairo]: Code updated

File: src/token/common/erc2981/erc2981.cairo (hunk 12 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo]: Code updated

File: src/tests/token/erc2981/test_erc2981.cairo (hunk 13 above)
Comment (ptisserand) [reply src/tests/token/erc2981/test_erc2981.cairo:86]: Tests added

File: src/tests/mocks/erc2981_mocks.cairo (hunk 14 above)
Comment (ptisserand) [reply src/tests/mocks/erc2981_mocks.cairo]: Ok, I have move this part

File: src/token/common/erc2981/erc2981.cairo [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+    use openzeppelin::introspection::src5::SRC5Component::InternalTrait as SRC5InternalTrait;
+    use openzeppelin::introspection::src5::SRC5Component::SRC5Impl;
+    use openzeppelin::introspection::src5::SRC5Component;
+    use openzeppelin::token::common::erc2981::interface::{IERC2981, IERC2981_ID};
+
+    use starknet::ContractAddress;
+    use starknet::storage::Map;
+
+    const default_fee_denominator: u256 = 10_000;
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:18]:
```suggestion
    const DEFAULT_FEE_DENOMINATOR: u256 = 10_000;
```

File: src/token/common/erc2981/erc2981.cairo [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    #[storage]
+    struct Storage {
+        default_royalty_info: RoyaltyInfo,
+        token_royalty_info: Map<u256, RoyaltyInfo>,
+    }
+
+    //
+    // External
+    //
+    #[embeddable_as(ERC2981Impl)]
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:35]:
```suggestion

    #[embeddable_as(ERC2981Impl)]
```

There should be a newline in-between

File: src/token/common/erc2981/erc2981.cairo [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+            let royalty_info = self.token_royalty_info.read(token_id);
+
+            let (royalty_receiver, royalty_fraction) = if royalty_info.receiver.is_zero() {
+                let default_info = self.default_royalty_info.read();
+                (default_info.receiver, default_info.royalty_fraction)
+            } else {
+                (royalty_info.receiver, royalty_info.royalty_fraction)
+            };
+
+            let royalty_amount: u256 = (sale_price * royalty_fraction) / self._fee_denominator();
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:56]:
```suggestion
            let royalty_amount = (sale_price * royalty_fraction) / self._fee_denominator();
```

File: src/token/common/erc2981/erc2981.cairo [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+            (royalty_receiver, royalty_amount)
+        }
+    }
+
+
+    //
+    // Internal
+    //
+    #[generate_trait]
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:66]:
```suggestion

    #[generate_trait]
```

File: src/token/common/erc2981/erc2981.cairo [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+            fee_numerator: u256,
+        ) {
+            let denominator = self._fee_denominator();
+            assert!(fee_numerator <= denominator, "Invalid default royalty");
+            assert!(receiver.is_non_zero(), "Invalid default royalty receiver");
+            self
+                .default_royalty_info
+                .write(RoyaltyInfo { receiver, royalty_fraction: fee_numerator })
+        }
+
```
Comment (immrsd) [inline src/token/common/erc2981/erc2981.cairo:119]: This implementation lacks `_deleteDefaultRoyalty` function, which is present in the Solidity implementation. Imo, it makes sense to add it for consistency

File: src/token/common/erc2981/interface.cairo [hunk 20]
```diff
@@ -0,0 +1,8 @@
+use starknet::ContractAddress;
+
+pub const IERC2981_ID: felt252 = 0x2d3414e45a8700c29f119a54b9f11dca0e29e06ddcb214018fc37340e165ed6;
```
Comment (immrsd) [inline src/token/common/erc2981/interface.cairo:3]: We should also consider adding the interface ID that is specified in the [EIP-2981](https://eips.ethereum.org/EIPS/eip-2981) and in the Solidity implementation (`0x2a55205a`)

**Review by immrsd — CHANGES_REQUESTED** [review]

File: src/token/common/erc2981/erc2981.cairo (hunk 19 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo:119]: Could you please confirm if the following behavior is expected after `_deleteDefaultRoyalty` is called ?:
- Default Royalty is 0% and receiver address is address(0)

File: src/token/common/erc2981/interface.cairo (hunk 20 above)
Comment (ptisserand) [reply src/token/common/erc2981/interface.cairo:3]: Sorry I don't understand what you mean with Solidity implementation interface ID:
- Do you mean I need to add camelCase method `royaltyInfo` ?
- If yes, should IERC2981_ID be computed with this interface added or should I add additional interface ID ?

File: src/token/common/erc2981/erc2981.cairo (hunk 15 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo]: Code updated

File: src/token/common/erc2981/erc2981.cairo (hunk 16 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo:36]: Code updated

File: src/token/common/erc2981/erc2981.cairo (hunk 17 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo]: Code updated

File: src/token/common/erc2981/erc2981.cairo (hunk 18 above)
Comment (ptisserand) [reply src/token/common/erc2981/erc2981.cairo:68]: Code updated

File: src/token/common/erc2981/interface.cairo (hunk 20 above)
Comment (immrsd) [reply src/token/common/erc2981/interface.cairo:3]: We have discussed the issue internally and decided that we're not going to add the support for the interface IDs from Solidity

File: src/token/common/erc2981/erc2981.cairo (hunk 19 above)
Comment (immrsd) [reply src/token/common/erc2981/erc2981.cairo:121]: Yes, it is the expected one

**Comment by ericnordelo** [comment]
Thanks @ptisserand. Closing this in favor of #1091.
